attorney_online/include/datatypes.h

```diff
@@ -0,0 +1,541 @@
+      EVIDENCE_ID,
+      FLIP,
+      REALIZATION,
+      COLOR,
+      SHOWNAME,
+      OTHER_CHARID,
+      OTHER_NAME,
+      OTHER_EMOTE,
+      SELF_OFFSET,
+      OTHER_OFFSET,
+      OTHER_FLIP,
+      IMMEDIATE,
+      LOOPING_SFX,
+      SCREENSHAKE,
+      FRAME_SCREENSHAKE,
+      FRAME_REALIZATION,
+      FRAME_SFX,
+      ADDITIVE,
+      EFFECTS,
+    };
+
+    /**
+     * @brief Constructor for an incoming IC-Chat packet
+     *
+     * @details Due to the way AO messages are encoded, there is no way to
+     * safely determine what contents are in the packet without doing overly
+     * many size checks.
+     * So you get a KFO and CCCC packet extension check.
+     * If you decide to not send your arguments correctly, that's on you.
+     */
+    MSPacket(QStringList packet, bool cccc_features, bool kfo_features)
+    {
+      desk_mod = toDataType<DESKMOD>(packet.at(DESK_MOD));
+      pre_animation = packet.at(PRE_EMOTE);
+      character = packet.at(CHAR_NAME);
+      emote = packet.at(EMOTE);
+      message = packet.at(MESSAGE);
+      side = packet.at(SIDE);
+      sfx_name = packet.at(SFX_NAME);
+      emote_mod = toDataType<EMOTE_MODIFIER>(packet.at(EMOTE_MOD));
+      char_id = packet.at(CHAR_ID).toInt();
+      sfx_delay = packet.at(SFX_DELAY).toInt();
+      shout_mod = toDataType<SHOUT_MODIFIER>(packet.at(OBJECTION_MOD));
+      if (shout_mod != SHOUT_MODIFIER::NOTHING &&
+          emote_mod == EMOTE_MODIFIER::NONE) {
+        emote_mod = EMOTE_MODIFIER::OBJECTION;
+      }
+      evidence = packet.at(EVIDENCE_ID).toInt();
+      flip = toDataType<bool>(packet.at(FLIP));
+      realization = toDataType<bool>(packet.at(REALIZATION));
+      text_color = toDataType<TEXT_COLOR>(packet.at(COLOR));
+      if (text_color == TEXT_COLOR::RAINBOW) {
+        text_color = TEXT_COLOR::WHITE;
+      }
+
+      if (cccc_features) {
+        showname = packet.at(SHOWNAME);
+        other_char_id = packet.at(OTHER_CHARID).toInt();
+        other_char_name = packet.at(OTHER_NAME);
+        other_char_emote = packet.at(OTHER_EMOTE);
+
+        QStringList offset = packet.at(SELF_OFFSET).split("&");
+        if (offset.size() == 2) {
+          character_offset = {offset[0].toInt(), offset[1].toInt()};
+        }
+        else {
+          character_offset = {offset[0].toInt(), 0};
+        }
+
+        offset = packet.at(OTHER_OFFSET).split("&");
+        if (offset.size() == 2) {
+          other_offset = {offset[0].toInt(), offset[1].toInt()};
+        }
+        else {
+          other_offset = {offset[0].toInt(), 0};
+        }
+        other_flip = toDataType<bool>(packet.at(OTHER_FLIP));
+        immediate = toDataType<bool>(packet.at(IMMEDIATE));
+      }
+      if (kfo_features) {
+        sfx_looping = toDataType<bool>(packet.at(LOOPING_SFX));
+        screenshake = toDataType<bool>(packet.at(SCREENSHAKE));
+        if (!packet.at(FRAME_SFX).isEmpty() &&
+            Options().networkedFrameSfxEnabled()) {
```

Comment:
this looks smelly to be accessing settings from a struct. For instance, when you're saving messages to a replay, and you have networked frame sfx disabled, you might want to preserve that message data instead of throwing it away just because a clientside option was set.

Reply:
Valid, I forgot demo files were a thing while writing that part and that I need to actually handle them at some point, but rn I am more concerned with getting the fundamentals more right than wrong than worry about AOs more than few crimes against sane codebases.